---
title: "aquaponics microbiome"
author: "Demitri"
date: "4/3/2020"
output: 
  html_document:
    code_folding: hide
bibliography: references.bib
link-citations: yes  
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, message = FALSE, warning = FALSE, results = FALSE)
knitr::opts_knit$set(root.dir = "C:/Users/dlifg/Desktop/aquaponics sequence data/QIIME files/qiime2/")
```
## Import

get qiime files into R. I use *Phyloseq* to 

- clean up irregularites in taxonomy table
- remove ASVs associated with mitochondria and chloroplast
- rename ASV (e.g. ASV1, ASV2, etc.)
- make transformation to ratios
- normalize via rarefaction


```{r}
library("biomformat")
library("phyloseq")
library("ggplot2")
library("Biostrings")
library("tidyverse")
library("gridExtra")
library("vegan")

tree = "tree.nwk"
map = "meta.txt"
repseq = "dna-sequences.fasta"
biom = read_biom("table-with-taxonomy.biom")
data = import_biom(biom, tree, parseFunction = parse_taxonomy_greengenes)

#update and fix sample_data
meta = import_qiime_sample_data(map)
sample_data(data) = meta

#reference sequences
bs = readDNAStringSet(repseq)
data = merge_phyloseq(data, bs)

#remove mistaken taxanomic ranks
tax_table(data) <- tax_table(data)[, 1:7]

#remove chloroplasts & mitochondria
data = subset_taxa(data, Family != "Mitochondria" & Class != "Chloroplast" & Family != "mitochondria" )

#rename OTU names
taxa_names(data) = paste0("ASV",seq(ntaxa(data)))

#transform abundance to ratio
data_r = transform_sample_counts(data, function(x) 100* x/sum(x))

#rarefy to min. sequence count
rngseed1=100
data_rarefy = rarefy_even_depth(data, sample.size = min(sample_sums(data)), rngseed = rngseed1, replace = TRUE, trimOTUs = TRUE, verbose = TRUE)
data_rarefy_r = transform_sample_counts(data_rarefy, function(x) 100* x/sum(x))

```
```{r}
sequence_sum = sample_sums(data)
otu_sum = apply(otu_table(data)>0, 2, sum)
sample_info = as.data.frame(cbind(as(sample_data(data), "matrix"), sequence_sum, otu_sum)) %>%
  select(Tank, Type, sequence_sum, otu_sum)%>%
  rename(Sequences = sequence_sum, ASVs = otu_sum)

#reorder rows
slist = paste0(rep("sample-", times = 16), seq.int(1,16), "")
sample_info = sample_info[slist,]
```

There was a total of `r prettyNum(sum(sequence_sum), big.mark = ",")` quality filtered reads and `r ntaxa(data)` ASVs in the data set. The average reads per sample was `r prettyNum(mean(sequence_sum), big.mark = ",")` The lowest count was `r prettyNum(min(sequence_sum), big.mark = ",")` in sample 10 and the highest was `r prettyNum(max(sequence_sum), big.mark = ",")` in sample 1. An overview of the samples is shown in the following table.

```{r echo = FALSE, results = 'asis'}
library(rmarkdown)
paged_table(sample_info)

```


## Rarefaction

Rarefaction curves were produced using *vegan* function *rarecurve*. Each curve shows the number of ASVs or bacteria species found in a sample at different sequence counts. As the number of 16S sequences increases, the slope of all curves approach zero. This indicates that the sampling depth was satisfactory and the samples are a fair representation of the bacteria community

```{r}
library(vegan)

#function to transpose phyloseq OTU table into vegan OTU matrix 
veganotu = function(physeq) {
  require("vegan")
  OTU = otu_table(physeq)
  if (taxa_are_rows(OTU)) {
    OTU = t(OTU)
  }
  return(as(OTU, "matrix"))
}

#transpose phyloseq table to vegan-compatible matrix 
rcurve_data = veganotu(data)

#xlimit at median sequences
r = rarecurve(rcurve_data, step = 20, xlab = "16S Sequences", ylab = "ASVs", label = FALSE, xlim = c(0,17633))
```

## Alpha Diversity
Richness and Eveness were used to compare the diversity within samples. Richness refers to the number of ASVs or species and eveness to how they are distrbuted. Pielouâ€™s eveness index $$J=H/ln(S)$$ was derived using *vegan* function *diversity*. 

```{r cache=TRUE}
#function to transpose phyloseq OTU table into vegan OTU matrix 
veganotu = function(physeq) {
  require("vegan")
  OTU = otu_table(physeq)
  if (taxa_are_rows(OTU)) {
    OTU = t(OTU)
  }
  return(as(OTU, "matrix"))
}

#use rarefied counts
v_data = veganotu(data_rarefy)

#get sum of ASV's for each sample
rich = specnumber(v_data)

#Shannon diversity score
H = diversity(v_data, index = "shannon")

#Pielous's Eveness
J <- H/log(specnumber(v_data))

#make table 
x = as.data.frame(cbind(as(sample_data(data), "matrix"), H, J, rich))
x$H = as.numeric(as.character(x$H))
x$J = as.numeric(as.character(x$J))
x$rich = as.numeric(as.character(x$rich))


p = ggplot(x, aes(x = Type, y = rich))+
  geom_boxplot(coef =4, alpha = 0.5)+
  geom_jitter(color = "black", size = 0.6, alpha = 0.9)+
  ylab("Richness")+
  xlab("")

pp = ggplot(x, aes(x = Type, y = H))+
  stat_boxplot(coef = 4, alpha = 0.5)+
  geom_jitter(color = "black", size = 0.6, alpha = 0.9)+
  ylab("Pielous's Evenness")+
  xlab("")
grid.arrange(p,pp, nrow = 1)
```

Overall, the substrate environment had greater diversity than the water environment. The gut was signficantly less diverse than water and substrate enivornments. 


## Beta Diversity

### Hierarchical cluster analysis

Hierarachcical clustering of Bray-Curtis dissimilarity matrix using Average-linkage method from package *hclust* shows clustering of samples by Type. Substrate samples from the same tank form discrete clusters while water samples are more ambigious. 

```{r}
source('C:/Users/dlifg/Desktop/aquaponics sequence data/QIIME files/qiime2/scripts/plot_dendrogram.R')
plot_dendrogram(data_rarefy,"average", "Tank")
```

### Non-metric dimensional scaling (NMDS)

NMDS ordination of Bray-Curtis distance using *vegan* function *metaMDS* with dimensions, k = 5 and stress = 0.012. Ellipses 
validates hierarchical clustery of samples by Type. 

```{r}
#convert phyloseq sample data into data frame
v_meta = as.data.frame(as.matrix(sample_data(data_rarefy)))

#picked 5 dimenstions (k = 5) as stress 
bc.nmds = metaMDS(v_data, distance = "bray", k = 5, trymax = 100) 

par(mar=c(5.1, 4.1, 4.1, 8.1), xpd=TRUE)
plot(bc.nmds, type = "n")
points(bc.nmds, display = "sites", col=c("grey45","darkred", "darkblue")[v_meta$Type])
ordiellipse(bc.nmds, groups = v_meta$Type, kind = "se", conf = .99, draw = "lines", col = c("grey45","darkred", "darkblue"), label = TRUE, alpha = 100)

#legend("topright", legend = c("gut", "substrate", "water"), col = c("grey45","darkred", "darkblue"), 
#       pch=20, inset = c(-0.2,0), title = "Sample Type", y.intersp = 1, x.intersp = 1, bty = "n", pt.cex = 2)
```

### Adonis

```{r results = TRUE }
doadonis <- function(physeq, category, distance) {
  bdist <- phyloseq::distance(physeq, distance)
  col <- as(sample_data(physeq), "data.frame")[ ,category]
  
  # Adonis test -- do the centroid differ according to category (i.e sampele type)
  adonis.bdist <- adonis(bdist ~ col)
  print("Adonis results:")
  print(adonis.bdist)
  
  # Homogeneity of dispersion test -- does the variance of the communties differ according to category (i.e. sample type)
  betatax = betadisper(bdist,col)
  p = permutest(betatax)
  print("Betadisper results:")
  print(p$tab)
}

doadonis(data_rarefy, "Type", "bray")
```

## Composition

### Phylum-level

Bar plots showing the abundance of phyla was generated in *ggplot*. Any phyla contributing less than 1% of total abundance in any one sample were grouped together in category: **<1%**.

```{r fig.height = 6}
source('C:/Users/dlifg/Desktop/aquaponics sequence data/QIIME files/qiime2/scripts/plot_phylum.R')
plot_phylum(data_rarefy_r)
```

In total, 15 different phyla were identified with 11 contributing over 99% of the abundance in any given sample. Bacteriodetes, Proteobacteria, Fusobacteria, Verrucomicrobia, and Actinobacteria were the most prevelant in all samples but differences were seen in the different environment types. Proteobacteria made up  only 1%-2% of the gut community but were enriched in water and substrate environments making up between 22% - 44% of the community. Actinobacteria was prevelent in the water environment (19% average abundance) but much less so in substrate enivornment (2% average abundance). Planctomycetes, Chlorflexi, and Acidobacteria on the other hand were more prevelant in the substrate environment. Only Nitrospirae showed any tank specific localization, as it was only prevelant in the biofilter substrate site. 

### ASV-level

Heatmat generated with *heatmap.2* shows the general composition of the system. Only ASV's making up at least 1% of the abundance in any single sample are included. Samples (columns) are ordered by clustering of Bray-Curtis dissimilarity of the samples entire community. ASVs (rows) are ordered by Euclidean distance based clustering. Phylum and the highest taxonomic level is shown for each ASV. The number in brackets indicate the number of ASVs with identical taxonomy.

```{r fig.height= 10}
source('C:/Users/dlifg/Desktop/aquaponics sequence data/QIIME files/qiime2/scripts/ASV_heatmap.R')
hmap(data_rarefy_r,1)
```

Looking at the community at a higher taxonomic resolution shows a clearer picture of which taxa are different in the gut, water, and substrate environments. The gut was dominanted by species Cetobacterium somerae, Akkermansia muciniphila and genre Paludibacter, Parabacteroides, all of which are known to be anaerobic -- a good indicator that these are true gut microbes. C.somerae has been well documented in the gut of freshwater fish such as Nile tilpia **@Tsuchiya2008a** and has been identified in aquaponic systems **@Schmautz2017**. It was abundant in the water environment throughout our system ranging from 4% in the hydroponic tank to 16% in the fish tank. This falls in line with another study which found C. somerae ubiquitous in the water environment in multipe distinct RAS/aquaponic facilities **@Bartelme2019**. In addition, Paludibacter and A. muciniphila were abundant throughout the system which strongly indicates that microbes dervied from the fish gut are a steady presence in minimal mechanical filtration aquaponics.  
The water envrionment was mostly enriched with common bacteria found in freshwater lakes and rivers such as genre Polynucleobacter, Limnohabitans, Sediminibacterium, and ACK-M1 **@Zwart2002**.  A single ASV classified to ACK-M1 made up over 20% of the water community in the fish tank, sump, and hydroponic tank. Genus Novosphingobium, a metabolically versatile aerobic bacteria that has been isolated from a wide range of habitats **Kumar2017**, was relateively abundant in all water enivronments (1% - 11%). There was a significant presence of heterotrophic bacteria in the water such as genre Flavobacterium and Mycobacterium, 


## ASV's driving the the dissimlarity between aquaponic environments

Based on the previous results, the water community was si . Clearly certain ASVs were more abundant in one environment type than another. To get better understanding of which ASVs were causing th 
In an attempt to discover which ASVs are causing the water, substrate, and gut communities to differ, I tried out several methods.


### Method 1: DESeq2 differential abundance

*DESeq2* package provides methods to test for differential expression by use of negative binomial models. Orignally implemented for RNA-Seq, it has been adopted for microbiome analyses in *phyloseq*. 
I followed this phyloseq tutorial: https://bioconductor.org/packages/devel/bioc/vignettes/phyloseq/inst/doc/phyloseq-mixture-models.html 

```{r}
library(DESeq2)

#use unrarefied data 
#remove gut samples because I'm intersted in difference between water and substrate environment
#low number of gut samples could skew test
data_noGut = subset_samples(data, Type != "gut")

#convert to deseq2 class using Type as study design factor. Type refers to either water or substrate 
data_deseq = phyloseq_to_deseq2(data_noGut, ~ Type)

#calculate modified geometric means that ignores zeroes. 
gm_mean = function(x, na.rm=TRUE){
  exp(sum(log(x[x > 0]), na.rm=na.rm) / length(x))
}
geoMeans = apply(counts(data_deseq), 1, gm_mean)

#calculate normalize factors using modified geometric means
data_deseq = estimateSizeFactors(data_deseq, geoMeans = geoMeans)

#performs differential expression based on the negative binomial distribution. Used default test (Wald). fitType = "local" for dispersion estimates
data_deseq = DESeq(data_deseq, fitType="local")

#get results of test
res = results(data_deseq, cooksCutoff = FALSE)
res = res[order(res$padj, na.last=NA), ]

#select for P value <0.01
alpha = 0.01
sigtab = res[(res$padj < alpha), ]

#add taxonmy to results table
sigtab = cbind(as(sigtab, "data.frame"), as(tax_table(data_noGut)[rownames(sigtab), ], "matrix"))
sigtab = sigtab[, c("baseMean", "log2FoldChange", "padj", "Phylum", "Class", "Family", "Genus", "Species")]

```

The following table shows the results of the differntial abunance test. ASV's that 
```{r echo=FALSE, results = 'asis'}
library(rmarkdown)
paged_table(sigtab)

```

```{r}
# Phylum order
x = tapply(sigtab$log2FoldChange, sigtab$Phylum, function(x) max(x))
x = sort(x, TRUE)
sigtab$Phylum = factor(as.character(sigtab$Phylum), levels=names(x))
# Genus order
x = tapply(sigtab$log2FoldChange, sigtab$Genus, function(x) max(x))
x = sort(x, TRUE)
sigtab$Genus = factor(as.character(sigtab$Genus), levels=names(x))
#plot
ggplot(sigtab, aes(x=Genus, y=log2FoldChange, color=Phylum)) + geom_point(size=3) + 
  theme(axis.text.x = element_text(angle = -90, hjust = 0, vjust=0.5))+
  scale_color_manual(values = phylum_colors2)
```


### Method 2: SIMPER

*Vegan* function *SIMPER* assesses the average percent contribution of individual ASV's to the Bracy-Curtis dissimilarity of sample types. It answer's the question: which ASV are contributing most to the differencce in water and substrate environments. H

```{r}
#used on 
simp = simper(v_data, v_meta$Type, permutations=100)
simp_sum = summary(simp)

water_substrate = simp_sum$water_substrate[1:10,]
water_substrate =  cbind(water_substrate, as(tax_table(data)[rownames(water_substrate), ], "matrix"))

```
```{r echo=FALSE, results = 'asis'}
library(rmarkdown)
paged_table(water_substrate)
```
```{r}

```

